# Como manipular dados textuais

Inicializar os pacotes e ler os arquivos

In [1]:
import pandas as pd

titanic = pd.read_csv("data/titanic.csv")
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Coloque todos os caracteres do nome em minúsculas.  

Para transformar cada uma das strings da coluna Name em minúsculas,  
selecione a coluna Name (consulte o tutorial sobre seleção de dados ),  
adicione o stracessador e aplique o lowermétodo.  
Dessa forma, cada uma das strings é convertida elemento a elemento.

In [2]:
titanic["Name"].str.lower()

0                                braund, mr. owen harris
1      cumings, mrs. john bradley (florence briggs th...
2                                  heikkinen, miss laina
3           futrelle, mrs. jacques heath (lily may peel)
4                               allen, mr. william henry
                             ...                        
886                                montvila, rev. juozas
887                          graham, miss margaret edith
888              johnston, miss catherine helen "carrie"
889                                behr, mr. karl howell
890                                  dooley, mr. patrick
Name: Name, Length: 891, dtype: object

Semelhante aos objetos datetime no tutorial de séries temporais que possuem um dt acessador,  
vários métodos de string especializados estão disponíveis ao usar o str acessador.  
Esses métodos geralmente têm nomes correspondentes aos métodos de string integrados equivalentes para elementos individuais,  
mas são aplicados elemento a elemento (lembra dos cálculos elemento a elemento ?) em cada um dos valores das colunas.

Crie uma nova coluna Surnameque contenha o sobrenome dos passageiros extraindo a parte antes da vírgula.

In [3]:
titanic["Name"].str.split(",")

0                             [Braund,  Mr. Owen Harris]
1      [Cumings,  Mrs. John Bradley (Florence Briggs ...
2                               [Heikkinen,  Miss Laina]
3        [Futrelle,  Mrs. Jacques Heath (Lily May Peel)]
4                            [Allen,  Mr. William Henry]
                             ...                        
886                             [Montvila,  Rev. Juozas]
887                       [Graham,  Miss Margaret Edith]
888           [Johnston,  Miss Catherine Helen "Carrie"]
889                             [Behr,  Mr. Karl Howell]
890                               [Dooley,  Mr. Patrick]
Name: Name, Length: 891, dtype: object

Usando o método *Series.str.split()*, cada um dos valores é retornado como uma lista de 2 elementos.  
O primeiro elemento é a parte antes da vírgula e o segundo elemento é a parte depois da vírgula.

Como estamos interessados ​​apenas na primeira parte que representa o sobrenome (elemento 0),  
podemos usar novamente o stracessador e aplicar Series.str.get() para extrair a parte relevante.  
De fato, essas funções de string podem ser concatenadas para combinar várias funções ao mesmo tempo!

In [4]:
titanic["Surname"] = titanic["Name"].str.split(",").str.get(0)
titanic["Surname"]

0         Braund
1        Cumings
2      Heikkinen
3       Futrelle
4          Allen
         ...    
886     Montvila
887       Graham
888     Johnston
889         Behr
890       Dooley
Name: Surname, Length: 891, dtype: object

Extraia os dados dos passageiros sobre as condessas a bordo do Titanic.

In [5]:
titanic["Name"].str.contains("Countess")

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Name, Length: 891, dtype: bool

In [6]:
titanic[titanic["Name"].str.contains("Countess")]

PassengerId  Survived  Pclass  \
759          760         1       1   

                                                  Name     Sex   Age  SibSp  \
759  Rothes, the Countess. of (Lucy Noel Martha Dye...  female  33.0      0   

     Parch  Ticket  Fare Cabin Embarked Surname  
759      0  110152  86.5   B77        S  Rothes

O método string Series.str.contains() verifica, para cada um dos valores na coluna, se Name contém a palavra *Countess* e retorna,  
para cada um dos valores True, (Countess faz parte do nome) ou False (Countess não faz parte do nome). 

Esta saída pode ser usada para subselecionar os dados usando a indexação condicional (booleana) introduzida no tutorial sobre subconjuntos de dados.  
Como havia apenas uma condessa no Titanic, obtemos uma linha como resultado.

Extrações mais poderosas em strings são suportadas, pois os métodos Series.str.contains() e Series.str.extract()  
aceitam expressões regulares , mas estão fora do escopo deste tutorial.

Qual passageiro do Titanic tem o nome mais longo?

In [7]:
titanic["Name"].str.len()

0      23
1      51
2      21
3      44
4      24
       ..
886    21
887    27
888    39
889    21
890    19
Name: Name, Length: 891, dtype: int64

Para obter o nome mais longo, primeiro precisamos obter o comprimento de cada um dos nomes na coluna Name.  
Usando os métodos de string do Pandas, a função *Series.str.len()* é aplicada a cada um dos nomes individualmente (elemento por elemento).

In [8]:
titanic["Name"].str.len().idxmax()

307

Em seguida, precisamos obter o local correspondente, de preferência o rótulo do índice, na tabela cujo comprimento do nome é o maior.  
O método *idxmax()*, faz exatamente isso. Não é um método de string e é aplicado a inteiros, portanto, "no" stré usado.

In [9]:
titanic.loc[titanic["Name"].str.len().idxmax(), "Name"]

'Penasco y Castellana, Mrs. Victor de Satode (Maria Josefa Perez de Soto y Vallejo)'

Com base no nome do índice da linha (307) e da coluna (Name),  
podemos fazer uma seleção usando o loc operador, introduzido no tutorial sobre subconjuntos.

Na coluna “Sexo”, substitua os valores de “masculino” por “M” e os valores de “feminino” por “F”.

In [10]:
titanic["Sex_short"] = titanic["Sex"].replace({"male": "M", "female": "F"})
titanic["Sex_short"]

0      M
1      F
2      F
3      F
4      M
      ..
886    M
887    F
888    F
889    M
890    M
Name: Sex_short, Length: 891, dtype: object

Embora *replace()* não seja um método de string, ele fornece uma maneira conveniente de usar mapeamentos ou vocabulários para traduzir determinados valores.  
Requer um dictionarypara definir o mapeamento .{from : to}

Também existe um métdodo *replace()*, disponível para substituir um conjunto específico de caracteres.  
No entanto, ao mapear vários valores, isso se tornaria:

```
titanic["Sex_short"] = titanic["Sex"].str.replace("female", "F")
titanic["Sex_short"] = titanic["Sex_short"].str.replace("male", "M")
```

Isso se tornaria trabalhoso e facilmente levaria a erros. 
Imagine (ou tente você mesmo) o que aconteceria se essas duas afirmações fossem aplicadas na ordem inversa...

**LEMBRAR**

* Os métodos de string estão disponíveis usando o stracessador.
* Os métodos de string funcionam por elemento e podem ser usados ​​para indexação condicional.
* O replacemétodo é um método conveniente para converter valores de acordo com um determinado dicionário.